src/interpreter/explicitControlEvaluator.ts

```diff
@@ -1,7 +1,286 @@
+import { constructMainCallExpression } from '../ast/constructors';
+import { InvalidFunctionApplication } from './errors';
+import {
+  constructGlobalEnvironment,
+  extendEnvironment,
+  getEnvironmentValue
+} from './environment';
+import { isEmptyStatement } from '../ast/typeGuards';
+
+const evaluators: AgendaItemEvaluatorMapping = {
+  AssignmentExpression: (
+    command: AssignmentExpression,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  BinaryExpression: (
+    command: BinaryExpression,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  BlockStatement: (
+    command: BlockStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  BreakStatement: (
+    command: BreakStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  CallExpression: (
+    command: CallExpression,
+    state: ExplicitControlEvaluatorState
+  ) => {
+    state.agenda.push(
+      constructFunctionApplicationInstr(command.arguments.length, command)
+    );
+    for (let index = 0; index < command.arguments.length; index--) {
+      state.agenda.push(command.arguments[index]);
+    }
+    state.agenda.push(command.id);
+  },
+  Constant: (command: Constant, state: ExplicitControlEvaluatorState) => {
+    state.stash.push(command);
+  },
+  ContinueStatement: (
+    command: ContinueStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  DefaultStatement: (
+    command: DefaultStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  DoWhileStatement: (
+    command: DoWhileStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  EmptyStatement: (
+    command: EmptyStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  Environment: (
+    command: EnvironmentInstr,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  ExpressionStatement: (
+    command: ExpressionStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  ForStatement: (
+    command: ForStatement,
+    state: ExplicitControlEvaluatorState
+  ) => {},
+  FunctionApplication: (
+    command: FunctionApplicationInstr,
+    state: ExplicitControlEvaluatorState
+  ) => {
+    // Args are in order of left to right
+    const args: Value[] = [];
+    for (let index = 0; index < command.numOfArgs; index++) {
+      args.push(state.stash.pop());
+    }
+
+    const functionAddress = getEnvironmentValue(
+      command.functionId.name,
+      state.environment
+    );
+    const closureIdx = state.memory.get(functionAddress);
+    const closure = state.textMemory.get(closureIdx);
+
+    if (closure.params.length !== args.length) {
+      throw new InvalidFunctionApplication(
+        `Function takes in ${closure.params.length} arguments but ${args.length} arguments were passed in`
+      );
+    }
+
+    if (isEmptyStatement(closure.body)) {
+      return;
+    }
+
+    // TODO: Handle Tail Call in future.
+    if (
+      state.agenda.size() === 0 ||
+      state.agenda.peek().type === 'Environment'
```

Comment:
Would it be better if the `EnvironmentInstr` were to be a `RestoreEnvironmentInstr` instead to be more expressive?